brushfire-core/src/main/scala/com/stripe/brushfire/local/Trainer.scala

```diff
@@ -3,81 +3,176 @@ package com.stripe.brushfire.local
 
-case class Trainer[K: Ordering, V, T: Monoid](
-    trainingData: Iterable[Instance[K, V, T]],
-    sampler: Sampler[K],
-    trees: List[Tree[K, V, T]]) {
 
-  private def updateTrees(fn: (Tree[K, V, T], Int, Map[LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]]) => Tree[K, V, T]): Trainer[K, V, T] = {
-    val newTrees = trees.zipWithIndex.par.map {
-      case (tree, index) =>
-        val byLeaf =
-          trainingData.flatMap { instance =>
-            val repeats = sampler.timesInTrainingSet(instance.id, instance.timestamp, index)
-            if (repeats > 0) {
-              tree.leafFor(instance.features).map { leaf =>
-                1.to(repeats).toList.map { i => (instance, leaf) }
-              }.getOrElse(Nil)
-            } else {
-              Nil
-            }
-          }.groupBy { _._2 }
-            .mapValues { _.map { _._1 } }
-        fn(tree, index, byLeaf)
-    }
-    copy(trees = newTrees.toList)
+trait TrainingStep[K,V,T] {
+
+  type K1
+  type V1
+  type V2
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]): Seq[((Int,Int,K1), V1)]
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: V1): Traversable[V2]
+  def update(tree: Tree[K,V,T], map: Map[Int,V2]): Tree[K,V,T]
+
+  def semigroup1: Semigroup[V1]
+  def semigroup2: Semigroup[V2]
+  def ordering: Ordering[K1]
+}
+
+case class UpdateTargets[K,V,T](sampler: Sampler[K])(implicit val semigroup1: Semigroup[T])
+  extends TrainingStep[K,V,T] {
+
+  type K1 = Unit
+  type V1 = T
+  type V2 = T
+
+  val semigroup2 = semigroup1
+  val ordering = implicitly[Ordering[Unit]]
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+     for (
+        (treeIndex, tree) <- trees.toList;
+        i <- 1.to(sampler.timesInTrainingSet(instance.id, instance.timestamp, treeIndex)).toList;
+        leafIndex <- tree.leafIndexFor(instance.features).toList
+      ) yield (treeIndex, leafIndex, ()) -> instance.target
   }
 
-  private def updateLeaves(fn: (Int, LeafNode[K, V, T, Unit], Iterable[Instance[K, V, T]]) => Node[K, V, T, Unit]): Trainer[K, V, T] = {
-    updateTrees {
-      case (tree, treeIndex, byLeaf) =>
-        val newNodes = byLeaf.map {
-          case (leaf, instances) =>
-            leaf.index -> fn(treeIndex, leaf, instances)
+  def lift(tree: Tree[K,V,T], leafIndex: Int, key: K1, v1: T) = List(v1)
+
+  def update(tree: Tree[K,V,T], map: Map[Int,T]) = {
+    tree.updateByLeafIndex { index => map.get(index).map { t => LeafNode(index, t) } }
+  }
+}
+
+case class Expand[K,V,T](sampler: Sampler[K], stopper: Stopper[T], splitter: Splitter[V, T], evaluator: Evaluator[V, T])(implicit val ordering: Ordering[K])
+  extends TrainingStep[K,V,T] {
+
+  type K1 = K
+  type V1 = splitter.S
+  type V2 = (K, Split[V, T], Double)
+
+  def prepare(trees: Map[Int, Tree[K,V,T]], instance: Instance[K,V,T]) = {
+    lazy val features = instance.features.mapValues { value => splitter.create(value, instance.target) }
```

Comment:
Minor nit - `mapValues` here will recompute the values each time they are accessed, which in our case is O(# trees + # times in training sets). It sucks, but would probably be best to use `map { case (key, value) => (key, splitter.create(value, instance.target)) }` :\